File: day1/task2.hs
import Data.List (sort)

readNums :: String -> (Int, Int)
readNums s =
  let (n, s') = head (reads s :: [(Int, String)])
      (n', _) = head (reads s' :: [(Int, String)])
   in (n, n')

readLines :: [String] -> ([Int], [Int])
readLines [] = ([], [])
readLines (x : xs) =
  let (n, n') = readNums x
      (ns, ns') = readLines xs
   in (n : ns, n' : ns')

shrinkRec :: [Int] -> Int -> Int -> [(Int, Int)]
shrinkRec [] n c = [(n, c)]
shrinkRec (x : xs) n c =
  if x == n
    then shrinkRec xs n (c + 1)
    else (n, c) : shrinkRec xs x 1

shrink :: [Int] -> [(Int, Int)]
shrink [] = []
shrink (x : xs) = shrinkRec (x : xs) x 0

similarity :: [(Int, Int)] -> [(Int, Int)] -> Int
similarity [] _ = 0
similarity _ [] = 0
similarity ((x, a) : xs) ((x', a') : xs')
  | x == x' = x * a * a' + similarity xs xs'
  | x > x' = similarity ((x, a) : xs) xs'
  | otherwise = similarity xs ((x', a') : xs')

main = do
  contents <- readFile "input.txt"
  let (l, l') = readLines $ lines contents
  print $ similarity (shrink $ sort l) (shrink $ sort l')

File: day1/task2.py
with open("input.txt", "r") as f:
    nums = [l.split("   ") for l in f.readlines()]

l1 = [int(n) for n, _ in nums]
l2 = [int(n) for _, n in nums]

m1, m2 = {}, {}
for n in l1:
    m1[n] = m1.get(n, 0) + 1
for n in l2:
    m2[n] = m2.get(n, 0) + 1

res = sum(k * m1[k] * m2.get(k, 0) for k in m1.keys())
print(res)
